### ISSUE TITLE
tf.keras model.evaluate (and fit with validation) possible leak when running on TPU

### ISSUE THREAD

**Issue by raulmt** [tensorflow/tensorflow#32265]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **'Linux-4.14.137+-x86_64-with-Ubuntu-18.04-bionic'** (Google Colab)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **No**
- TensorFlow installed from (source or binary): **binary** (I think, it's the one already installed in Colab)
- TensorFlow version (use command below): **1.14.0**
- Python version: **3.6.8**
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory: **Google Colab's TPU**

**Describe the current behavior**
When running `model.evaluate` in TPU, each time it gets executed it takes longer and longer. I started seeing this behavior when training (`model.fit()`) but only when providing validation data. I guess `model.evaluate` gets called when processing the validation set, so I tried to have a minimum reproduction sample with just directly calling `model.evaluate`.

But please note this is also a problem when running `fit` with validation data. Each epoch takes longer and longer. I can start with 10 seconds on epoch 1 and have like 160 seconds on epoch 50.

Other *strange* and related behavior I can see:
- Just when calling `evaluate`, you get two identical lines as output instead of just 1 as I would expect (calling `fit` without validation data outputs 1 line per epoch, but when calling it with validation data it outputs that same 1 line plus two identical lines corresponding to the `evaluation`)
- each time `evaluate` is called it seems to incur in a "setup" time, since the step itself takes considerably less than than the total `evaluate` execution time
- as you call it again and again, it seems both, this "setup time" and the "evaluation time" itself increase, but not exactly as much. It seems the setup time increases more than the evaluation time
- the reproduction example below has very simple data and model, and you can still see this behavior with that. But it seems that with more complex models the time it takes for `evaluate` increases at a higher rate each time it gets called

**Describe the expected behavior**
Each time evaluate is called with the same data, it should take roughly the same time. Also, when running fit with validation, each epoch (after the first one at least) should also take roughly the same time (this does happen without validation data, but doesn't happen with validation)

**Code to reproduce the issue**

I tried to create a minimal example, so I'm not even training the model here before evaluating some data. But note this is also a problem when training, so I don't think using an untrained model here is related to the issue. In the same way, I'm using randomly generated data, but this happened to me when training with real images and masks, and also with higher batch size (I was using the recommended 1024, but I used 128 here so the example executes quicker)

```python
import os
import time
import numpy as np
import tensorflow as tf

tf.keras.backend.clear_session()

TF_MASTER = 'grpc://{}'.format(os.environ['COLAB_TPU_ADDR'])

with tf.Session(TF_MASTER) as session:
  print(session.list_devices())

print("Tensorflow Version:", tf.VERSION)
print("Tensorflow Keras Version:", tf.keras.__version__)

amount = 128
size = [256, 256]
images = np.array([np.random.rand(*size, 3).astype('float32') for i in range(amount)])
masks = np.array([np.random.rand(*size, 1).astype('float32') for i in range(amount)])

ds = tf.data.Dataset.from_tensor_slices((images, masks)).batch(amount, drop_remainder=True) # we need Dataset to run on TPU
print(ds)

# very minimal model to demostrate the issue
def make_model(batch_size=None):
  src = tf.keras.layers.Input(shape=(*size,3), batch_size=batch_size, dtype=tf.float32, name='Input')
  outputs = tf.keras.layers.Conv2D(1, (1, 1), activation='sigmoid')(src)

  model = tf.keras.Model(inputs=[src], outputs=[outputs])
  model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
  return model

resolver = tf.contrib.cluster_resolver.TPUClusterResolver(tpu='grpc://' + os.environ['COLAB_TPU_ADDR'])
tf.contrib.distribute.initialize_tpu_system(resolver)
strategy = tf.contrib.distribute.TPUStrategy(resolver)
with strategy.scope():
  model = make_model(batch_size = 128)
model.summary()

for i in range(30):
  start_time = time.time()
  model.evaluate(ds, steps=1)
  print("--- %s seconds ---" % (time.time() - start_time))
```

Output:
```
1/1 [==============================] - 2s 2s/step
1/1 [==============================] - 2s 2s/step
--- 4.957882404327393 seconds ---
1/1 [==============================] - 2s 2s/step
1/1 [==============================] - 2s 2s/step
--- 4.556092977523804 seconds ---
1/1 [==============================] - 2s 2s/step
1/1 [==============================] - 2s 2s/step
--- 4.7304792404174805 seconds ---
1/1 [==============================] - 2s 2s/step
1/1 [==============================] - 2s 2s/step
--- 5.154953479766846 seconds ---
1/1 [==============================] - 3s 3s/step
1/1 [==============================] - 3s 3s/step
--- 5.432474136352539 seconds ---
1/1 [==============================] - 3s 3s/step
1/1 [==============================] - 3s 3s/step
--- 5.930710315704346 seconds ---
1/1 [==============================] - 3s 3s/step
1/1 [==============================] - 3s 3s/step
--- 5.949801683425903 seconds ---
1/1 [==============================] - 3s 3s/step
1/1 [==============================] - 3s 3s/step
--- 6.352793455123901 seconds ---
1/1 [==============================] - 3s 3s/step
1/1 [==============================] - 3s 3s/step
--- 7.153834581375122 seconds ---
1/1 [==============================] - 4s 4s/step
1/1 [==============================] - 4s 4s/step
--- 7.319533348083496 seconds ---
1/1 [==============================] - 4s 4s/step
1/1 [==============================] - 4s 4s/step
--- 7.719737529754639 seconds ---
1/1 [==============================] - 4s 4s/step
1/1 [==============================] - 4s 4s/step
--- 8.089992761611938 seconds ---
1/1 [==============================] - 4s 4s/step
1/1 [==============================] - 4s 4s/step
--- 8.87194037437439 seconds ---
1/1 [==============================] - 5s 5s/step
1/1 [==============================] - 5s 5s/step
--- 9.02499771118164 seconds ---
1/1 [==============================] - 5s 5s/step
1/1 [==============================] - 5s 5s/step
--- 9.39006233215332 seconds ---
1/1 [==============================] - 5s 5s/step
1/1 [==============================] - 5s 5s/step
--- 9.687021970748901 seconds ---
1/1 [==============================] - 5s 5s/step
1/1 [==============================] - 5s 5s/step
--- 10.596904754638672 seconds ---
1/1 [==============================] - 6s 6s/step
1/1 [==============================] - 6s 6s/step
--- 10.745074272155762 seconds ---
1/1 [==============================] - 6s 6s/step
1/1 [==============================] - 6s 6s/step
--- 11.3403902053833 seconds ---
1/1 [==============================] - 6s 6s/step
1/1 [==============================] - 6s 6s/step
--- 11.945271015167236 seconds ---
1/1 [==============================] - 7s 7s/step
1/1 [==============================] - 7s 7s/step
--- 12.67944622039795 seconds ---
1/1 [==============================] - 7s 7s/step
1/1 [==============================] - 7s 7s/step
--- 13.195830345153809 seconds ---
1/1 [==============================] - 7s 7s/step
1/1 [==============================] - 7s 7s/step
--- 14.408772230148315 seconds ---
1/1 [==============================] - 7s 7s/step
1/1 [==============================] - 7s 7s/step
--- 14.38786506652832 seconds ---
1/1 [==============================] - 8s 8s/step
1/1 [==============================] - 8s 8s/step
--- 14.985581636428833 seconds ---
1/1 [==============================] - 8s 8s/step
1/1 [==============================] - 8s 8s/step
--- 15.89274287223816 seconds ---
1/1 [==============================] - 9s 9s/step
1/1 [==============================] - 9s 9s/step
--- 16.433870792388916 seconds ---
1/1 [==============================] - 9s 9s/step
1/1 [==============================] - 9s 9s/step
--- 17.08533024787903 seconds ---
[...]
```

**Other info / logs**

I created a Google Colab with this example, so you can execute it and check this behavior: https://colab.research.google.com/drive/1YgSSdlrfVpTfufGXNNXckYzOd-dzEbVI

**Comment by jvishnuvardhan** [comment]
I was able to reproduce the issue. Thanks!

**Comment by raulmt** [comment]
Btw, the workaround I'm using at the moment (since this is an important problem when training with validation data for more than a few epochs) is to use the `fit` arg `validation_freq`, with something like 5 or 10. That way you have many quick epochs (the ones without validation), and the evaluation stages you have are fewer and so they don't have the chance to increase their time enough to become a big problem. Of course, you loose intermediate validation, but at least you can quickly train 100 epochs without much trouble ;)

**Comment by jhseu** [comment]
Yeah, this is a known issue with the interaction with Keras + tf.distribute.Strategy in the TF 1.x. This is currently fixed in TF 2.x.

**Comment by raulmt** [comment]
@jhseu thanks, and do you know if latest 2.0 release works with cloud TPUs like in Colab? I read somewhere that it will be in 2.1 correct, with the TPUStrategy, is that correct? So the final solution will be to wait for 2.1 release in order to use TF with cloud TPU?

**Comment by jvishnuvardhan** [comment]
@raulmt TPUs are not officially supported in TF2.0. You can use `tf-nightly` which is `TF2.1` dev version. Thanks!

**Comment by raulmt** [comment]
Ok, thanks! Closing then…

**Comment by swghosh** [comment]
I'm still facing the same issue while on **TFv2.3** (and using a TPU). The model would train well but at the end of each epoch, the memory consumption (of the VM running the training) increases significantly for which we cannot train the model for longer durations.

PS: apart from `model.fit(...., validation_data=val_ds)`, I'm also using an extra validation set that calls `model.evaluate` with the help of the following callback like [this](https://github.com/tanzhenyu/image_augmentation/blob/master/image_augmentation/callbacks/extra_eval.py). I suspect that either of them should be responsible for the linear increase of memory consumption due to an internal memory leak!

Any suggested fixes would be of much help to me.

/cc: @jhseu, @tanzhenyu

**Comment by swghosh** [comment]
> I'm still facing the same issue while on **TFv2.3** (and using a TPU). The model would train well but at the end of each epoch, the memory consumption (of the VM running the training) increases significantly for which we cannot train the model for longer durations.
> 
> PS: apart from `model.fit(...., validation_data=val_ds)`, I'm also using an extra validation set that calls `model.evaluate` with the help of the following callback like [this](https://github.com/tanzhenyu/image_augmentation/blob/master/image_augmentation/callbacks/extra_eval.py). I suspect that either of them should be responsible for the linear increase of memory consumption due to an internal memory leak!
> 
> Any suggested fixes would be of much help to me.
> 
> /cc: @jhseu, @tanzhenyu

Here you could find two notebooks that would help prove the veracity of the claim. 
1. [TPU Experiment 1](https://colab.research.google.com/gist/swghosh/23f46273d82857c9cf8eb30423fcde4d/tpu-experiment-1.ipynb): Notebook demonstrating use of `model.fit` (on a TPU) with `validation_data` argument specified.
2. [TPU Experiment 2](https://colab.research.google.com/gist/swghosh/23f850cb447581d9f169a6914ed4681b/tpu-experiment-2.ipynb): Notebook demonstrating use of `model.fit` (on a TPU)
a) used with `validation_data` and an extra callback that does `model.evaluate(test_ds)` at `on_epoch_end`
b) used with `validation_data` only
c) used without `validation_data` argument.

Hope it helps.
